File: 백준/Gold/18405. 경쟁적 전염/경쟁적 전염.py
import sys
from collections import deque

# 입력
n, k = map(int, sys.stdin.readline().split())
graph = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
s, x, y = map(int, sys.stdin.readline().split())

# 상 하 좌 우
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

# 모든 바이러스 정보를 담는 리스트
viruses = []
for i in range(n):
    for j in range(n):
        if graph[i][j] != 0:
            # 바이러스 종류, 시간, 위치 X, 위치 Y를 추가
            viruses.append((graph[i][j], 0, i, j))

# 낮은 번호의 바이러스가 먼저 증식되도록 정렬
viruses.sort()
queue = deque(viruses)

while queue:
    virus, time, vx, vy = queue.popleft()
    if time == s:
        break
    for i in range(4):
        nx = vx + dx[i]
        ny = vy + dy[i]
        if 0 <= nx < n and 0 <= ny < n:
            if graph[nx][ny] == 0:
                graph[nx][ny] = virus
                queue.append((virus, time + 1, nx, ny))

print(graph[x-1][y-1])

File: 백준/Gold/18405. 경쟁적 전염/경쟁적 전염.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int[] dx = {0, 1, 0, -1};
        int[] dy = {1, 0, -1, 0};

        int n = Integer.parseInt(st.nextToken());;
        int k = Integer.parseInt(st.nextToken());;
        int[][] array = new int[n+1][n+1];
        ArrayList<Node> virus = new ArrayList<>();

        for(int i=1; i<=n; i++){
            st = new StringTokenizer(br.readLine());
            for(int j=1; j<=n; j++){
                array[i][j] = Integer.parseInt(st.nextToken());
                if (array[i][j] != 0) virus.add(new Node(array[i][j], 0, i, j));
            }
        }

        st = new StringTokenizer(br.readLine());
        int s = Integer.parseInt(st.nextToken());
        int x = Integer.parseInt(st.nextToken());
        int y = Integer.parseInt(st.nextToken());

        Collections.sort(virus);

        Queue<Node> queue = new LinkedList<>(virus);
        while(!queue.isEmpty()){
            Node poll = queue.poll();

            if (poll.time == s) break;
            for(int i=0; i<4; i++){
                int nx = poll.x + dx[i];
                int ny = poll.y + dy[i];

                if (nx >= 0 && nx <= n && ny >= 0 && ny <= n && array[nx][ny] == 0){
                    array[nx][ny] = poll.num;
                    queue.add(new Node(poll.num, poll.time + 1, nx, ny));
                }
            }
        }

        System.out.println(array[x][y]);
        br.close();
    }

    private static class Node implements Comparable<Node>{
        int num;
        int time;
        int x;
        int y;

        public Node(int num, int time, int x, int y) {
            this.num = num;
            this.time = time;
            this.x = x;
            this.y = y;
        }

        @Override
        public int compareTo(Node n){
            return Integer.compare(this.num, n.num);
        }
    }
}
